Cereal/CerealSerialization.swift

```diff
@@ -0,0 +1,434 @@
+        case let .StringValue(string):
+            buffer.append(CerealCoderTreeValueType.String.rawValue)
+            string.writeToBuffer(&buffer, stringMap: &stringMap)
+        case let .NSURLValue(url):
+            buffer.append(CerealCoderTreeValueType.NSURL.rawValue)
+            url.absoluteString.writeToBuffer(&buffer, stringMap: &stringMap)
+
+        default:
+            break
+        }
+    }
+
+    func writeIntBytes(inout buffer: [UInt8]) {
+        switch self {
+        case let .IntValue(int):
+            buffer.append(CerealCoderTreeValueType.Int.rawValue)
+            var array = toByteArray(int)
+            array.insertContentsOf(toByteArray(array.count), at: 0)
+            buffer.appendContentsOf(array)
+
+        case let .Int64Value(int64):
+            buffer.append(CerealCoderTreeValueType.Int64.rawValue)
+            var array = toByteArray(int64)
+            array.insertContentsOf(toByteArray(array.count), at: 0)
+            buffer.appendContentsOf(array)
+
+        case let .BoolValue(bool):
+            buffer.append(CerealCoderTreeValueType.Bool.rawValue)
+            buffer.appendContentsOf(toByteArray(1))
+            buffer.append(bool ? 1 : 0)
+
+        default:
+            break
+        }
+    }
+
+    func writeFloatBytes(inout buffer: [UInt8]) {
+        switch self {
+        case let .DoubleValue(double):
+            buffer.append(CerealCoderTreeValueType.Double.rawValue)
+            var array = toByteArray(double)
+            array.insertContentsOf(toByteArray(array.count), at: 0)
+            buffer.appendContentsOf(array)
+
+        case let .FloatValue(float):
+            buffer.append(CerealCoderTreeValueType.Float.rawValue)
+            var array = toByteArray(float)
+            array.insertContentsOf(toByteArray(array.count), at: 0)
+            buffer.appendContentsOf(array)
+
+        case let .NSDateValue(date):
+            let interval = date.timeIntervalSinceReferenceDate
+            buffer.append(CerealCoderTreeValueType.NSDate.rawValue)
+            var array = toByteArray(interval)
+            array.insertContentsOf(toByteArray(array.count), at: 0)
+            buffer.appendContentsOf(array)
+
+        default:
+            break
+        }
+    }
+}
+
+extension CoderTreeValue {
+    func toData() -> NSData {
+        let bytes = self.bytes()
+
+        return NSData(bytes: bytes, length: bytes.count)
+    }
+    func bytes() -> [UInt8] {
+        var result = [UInt8]()
+        result.reserveCapacity(20)
+
+        var stringMap = Dictionary<String, [UInt8]>(minimumCapacity: self.numberOfStringEntries())
+
+        self.writeToBuffer(&result, stringMap: &stringMap)
+        return result
+    }
+}
+
+// MARK: decoding extensions
+
+private extension String {
+    // http://stackoverflow.com/a/34100319/1254172 with a generic change
+    init<ByteSequence where ByteSequence: CollectionType, ByteSequence.Generator.Element == UInt8>(utf8stream: ByteSequence) {
```

Comment:
This is interesting. I think we don't need all this, though. Swift has a method that should handle this for us called `transcode`. I'm using it in another project of mine:

https://github.com/ketzusaka/Hummingbird/blob/master/Sources/String%2BTranscoding.swift